File: upstream.go
package main

import "database/sql"

type Upstream struct {
	ID           int
	Name         string
	UpstreamType string
	CreatedAt    int
	UpdatedAt    *int
}

func FindUpstreamByName(db *sql.DB, name string) (Upstream, error) {
	findStmt, err := db.Prepare("select id, name, upstream_id, created_at, updated_at from upstreams where name = ? limit 1")
	if err != nil {
		return Upstream{}, err
	}

	upstreamRow := findStmt.QueryRow(name)
	var upstream Upstream
	if err = upstreamRow.Scan(&upstream.ID, &upstream.Name, &upstream.UpstreamType, &upstream.CreatedAt, &upstream.UpdatedAt); err != nil {
		return Upstream{}, err
	}

	return upstream, nil
}

File: src/upstream.rs
use sqlx::{prelude::FromRow, SqliteConnection};

#[derive(FromRow)]
pub struct Upstream {
    pub id: i64,
    pub name: String,
    pub upstream_type: String,
    pub created_at: i64,
    pub updated_at: Option<i64>,
}

pub async fn find_upstream_by_name(
    conn: &mut SqliteConnection,
    name: &String,
) -> anyhow::Result<Option<Upstream>> {
    let upstream = sqlx::query_as!(Upstream, "select id, name, upstream_type, created_at, updated_at from upstreams where name = ? limit 1", name)
    .fetch_optional(conn).await?;

    Ok(upstream)
}
